<div style="border:solid green 2px; padding: 20px">
<b>Привет, Алексей!</b>

Меня зовут Александр Пономаренко, и я буду проверять твой проект. Предлагаю общаться на «ты» :) Но если это не удобно - дай знать, и мы перейдем на "вы". 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать data science. Ты уже проделал большую работу над проектом, но давай сделаем его еще лучше. Ниже ты найдешь мои комментарии - **пожалуйста, не перемещай, не изменяй и не удаляй их**. Увидев у тебя ошибку, в первый раз я лишь укажу на ее наличие и дам тебе возможность самой найти и исправить ее. На реальной работе твой начальник будет поступать так же, а я пытаюсь подготовить тебя именно к работе аналитиком. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку. Я буду использовать цветовую разметку:

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> Так выделены самые важные замечания. Без их отработки проект не будет принят. </div>

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Так выделены небольшие замечания. Я надеюсь, что их ты тоже учтешь - твой проект от этого станет только лучше. Но настаивать на их отработке не буду.

</div>

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Так я выделяю все остальные комментарии.</div>

Давай работать над проектом в диалоге: **если ты что-то меняешь в проекте или отвечаешь на мои комменатри — пиши об этом.** Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>Комментарий студента:</b> Например, вот так.</div>

Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп» с BERT

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 
Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.
Постройте модель со значением метрики качества F1 не меньше 0.75. 

In [1]:
!pip install transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.8/6.8 MB 51.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.6/7.6 MB 84.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 199.8/199.8 KB 15.8 MB/s eta 0:00:00


In [2]:
!pip install jupyter_server



Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 366.8/366.8 KB 7.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 80.6/80.6 KB 6.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 102.9/102.9 KB 11.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.9/55.9 KB 7.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 42.0 MB/s eta 0:00:00
  Attempting uninstall: pyzmq
    Found existing installation: pyzmq 23.2.1
    Uninstalling pyzmq-23.2.1:
      Successfully uninstalled pyzmq-23.2.1
  Attempting uninstall: jupyter-client
    Found existing installation: jupyter-client 6.1.12
    Uninstalling jupyter-client-6.1.12:
      Successfully uninstalled jupyter-client-6.1.12


In [3]:
!pip install catboost

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.6/76.6 MB 21.6 MB/s eta 0:00:00


## Импорт библиотек

In [4]:
import numpy as np
import pandas as pd
import torch
import transformers
import nltk
import warnings
warnings.filterwarnings("ignore")
import requests
import os

from tqdm import notebook
from tqdm.notebook import tqdm
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import ElasticNet
from sklearn.model_selection import train_test_split, StratifiedKFold, GridSearchCV
from transformers import AutoTokenizer, AutoModelForSequenceClassification
from torch.utils.data import TensorDataset
from catboost import CatBoostClassifier
from sklearn.metrics import f1_score
from sklearn.feature_extraction.text import TfidfVectorizer
from torch.utils.data import Dataset, DataLoader, TensorDataset
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
from sklearn.feature_extraction.text import TfidfVectorizer
from torch.nn.utils.rnn import pad_sequence
nltk.download('punkt')

# загрузка стоп-слов на английском языке
nltk.download('stopwords')
stopwords = set(stopwords.words('english'))

# загрузка модуля лемматизации слов из библиотеки nltk
nltk.download('wordnet')
lemmatizer = WordNetLemmatizer()
from sklearn.pipeline import Pipeline


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


In [5]:
df = pd.read_csv('https://code.s3.yandex.net/datasets/toxic_comments.csv', index_col=[0])


<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b>Да, хорошо, только лучше, чтобы и на платформе ЯП запускался код
    
    
```python
path_yandex = 'datasets/toxic_comments.csv'    
path = 'toxic_comments.csv'

if os.path.exists(path):
    data = pd.read_csv(path, index_col=[0])
    display(data.head(5))
elif os.path.exists(path_yandex):
    data = pd.read_csv(path_yandex, index_col=[0])
    display(data.head(5))
else:
    print('Ошибка в считывании данных')   
```

</div>

<div class="alert alert-info"> <b>Комментарий студента:</b> Поправил </div>


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 159292 entries, 0 to 159450
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   text    159292 non-null  object
 1   toxic   159292 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 3.6+ MB


In [7]:
df.isna().mean()

text     0.0
toxic    0.0
dtype: float64

## Подготовим выборки

In [8]:
df['toxic'].unique()

array([0, 1])

In [9]:
%%time

# функция для предобработки текста
def lemmatize_text(text):
    # токенизация
    tokens = word_tokenize(text.lower())
    
    # удаление стоп-слов и неалфавитных символов
    tokens = [t for t in tokens if t not in stopwords and t.isalpha()]
    
    # лемматизация
    lemmatized_tokens = [lemmatizer.lemmatize(t) for t in tokens]
    
    # объединение слов в строку и возврат
    return " ".join(lemmatized_tokens)

# предобработка текста
df['text'] = df['text'].apply(lemmatize_text)


CPU times: user 1min 48s, sys: 261 ms, total: 1min 48s
Wall time: 1min 47s


In [10]:
features = df.drop(columns=['toxic'], axis=1)
target = df['toxic']
features.shape, target.shape

((159292, 1), (159292,))

In [11]:
X_train, X_test, y_train, y_test = train_test_split(features['text'], target, test_size = 0.2, random_state=45) 
X_train.shape[0]/df.shape[0]

0.7999962333324963

In [12]:
# Вычисление весов классов
def class_weights(y):
    class_weights = {}
    n_classes = len(np.unique(y))
    bincount = np.bincount(y)
    n_samples = len(y)
    for i in range(n_classes):
        class_weights[i] = n_samples / (n_classes * bincount[i])
    return print("Weights:", class_weights,
                 "relation: {:.2f}".format(class_weights[1]/class_weights[0]))

In [13]:
class_weights(y_train)

Weights: {0: 0.5566023725911561, 1: 4.916775985801373} relation: 8.83


In [14]:
class_weights(y_test)

Weights: {0: 0.5563530315730651, 1: 4.936318562132011} relation: 8.87


Разделили выборки в соотношении 0.8:0.2: X_train:X_test. Вес класса в соотношении 1 к 8.83 для обучающей и 8.87 для тестовой

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Хорошо</div>

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> Перед tfidf следует данные обработать: удалить мусор + привести слова в нормальную форму (провести лемматизицаю)
    
+  https://webdevblog.ru/podhody-lemmatizacii-s-primerami-v-python/

</div>

In [15]:
df['text'].head(10)

0    explanation edits made username hardcore metal...
1    match background colour seemingly stuck thanks...
2    hey man really trying edit war guy constantly ...
3    ca make real suggestion improvement wondered s...
4                        sir hero chance remember page
5               congratulation well use tool well talk
6                          cocksucker piss around work
7    vandalism matt shirvington article reverted pl...
8    sorry word offensive anyway intending write an...
9                 alignment subject contrary dulithgow
Name: text, dtype: object

<div class="alert alert-info"> <b>Комментарий студента:</b> Добавил лемматизицаю перед разделением на выборки </div>


## Подберем лучшую модель

In [16]:
%%time

count_tf_idf = TfidfVectorizer()

# преобразование TF-IDF для тестовой выборки
X_train_tfidf = count_tf_idf.fit_transform(X_train)

# преобразование TF-IDF для тестовой выборки
X_test_tfidf = count_tf_idf.transform(X_test)

# вывод размеров матриц TF-IDF
print("Размер X_train_tfidf:", X_train_tfidf.shape)
print("Размер X_test_tfidf:", X_test_tfidf.shape)


Размер X_train_tfidf: (127433, 129078)
Размер X_test_tfidf: (31859, 129078)
CPU times: user 4.56 s, sys: 74.3 ms, total: 4.63 s
Wall time: 4.64 s


<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Отлично, молодец, верно используешь Tfidf.
    
    
Совет: Внутри кросс-валидации происходит разбиение выборки на train и valid. Однако, в таком случае векторизатор обучен на всей выборке(train), а это не совсем корректно. Чтобы избежать это можно воспользоваться Pipeline:
    
```python
pipeline = Pipeline([
    ('tfidf', TfidfVectorizer(stop_words=stopwords)),
    ('logreg', LogisticRegression(random_state=42)),
])
parameters = {
    'tfidf__max_df': (0.25, 0.5, 0.75),
    'tfidf__ngram_range': [(1, 1), (1, 2), (1, 3)],
    'logreg__C': [1,2,6]
}

grid_search_tune = RandomizedSearchCV(pipeline, parameters, cv=3, n_jobs=-1, scoring='f1', verbose=3)
grid_search_tune.fit(train_features, train_targets)
  
    
```
    
Это просто каркас, можешь сам выбрать какие параметры использовать для подбора:) 
    
+  https://runebook.dev/ru/docs/scikit_learn/modules/generated/sklearn.model_selection.halvinggridsearchcv - тут про HalvingGridSearchCV
    
+  https://www.rupython.com/python-sklearn-pipeline-pipeline-28301.html - про pipeline
+  https://towardsdatascience.com/how-to-use-sklearn-pipelines-for-ridiculously-neat-code-a61ab66ca90d

</div>

In [17]:
num_processors = os.cpu_count()
print(f"Количество процессоров (ядер): {num_processors}")


Количество процессоров (ядер): 12


In [21]:
%%time  
pipeline = Pipeline([  
    ('tfidf', TfidfVectorizer()),
    ('logreg', LogisticRegression(random_state=42)),
])  
  
parameters = {  
    'tfidf__max_df': (0.25, 0.5, 0.75),
    'logreg__C': [6, 10, 15, 17]
    }  

logreg_grid_search = GridSearchCV(pipeline, parameters, cv=3, n_jobs=-1,   scoring='f1')  
logreg_grid_search.fit(X_train, y_train)  
logreg_grid_search.best_estimator_

CPU times: user 25.4 s, sys: 1min 10s, total: 1min 36s
Wall time: 51.2 s


Pipeline(steps=[('tfidf', TfidfVectorizer(max_df=0.25)),
                ('logreg', LogisticRegression(C=15, random_state=42))])

In [22]:
print(f"Best params: {logreg_grid_search.best_params_}")
print(f'F1 лучшей модели {logreg_grid_search.best_score_}')

Best params: {'logreg__C': 15, 'tfidf__max_df': 0.25}
F1 лучшей модели 0.7723788061025139


In [23]:
%%time  
pipeline = Pipeline([  
    ('tfidf', TfidfVectorizer()),
    ('catboost', CatBoostClassifier(iterations=500, random_seed=42, loss_function='Logloss', task_type='GPU', verbose=False)),  
])  
  
parameters = {  
    'tfidf__max_df': (0.25, 0.5),
    'catboost__learning_rate': [0.03],  
    'catboost__depth': [6, 10],  
    'catboost__l2_leaf_reg': [1, 3],  
    'catboost__class_weights': [[1, 8.87]]
  }  

cat_grid_search = GridSearchCV(pipeline, parameters, cv=3, n_jobs=1,   scoring='f1')  
cat_grid_search.fit(X_train, y_train)  
cat_grid_search.best_estimator_

CPU times: user 2h 2min 24s, sys: 19min 14s, total: 2h 21min 38s
Wall time: 39min 46s


Pipeline(steps=[('tfidf', TfidfVectorizer(max_df=0.25)),
                ('catboost',
                 <catboost.core.CatBoostClassifier object at 0x7f3436dd9c10>)])

In [24]:
print(f"Best params: {cat_grid_search.best_params_}")
print(f'F1 лучшей модели {-cat_grid_search.best_score_}')
cat_rmse = (cat_grid_search.best_score_)

Best params: {'catboost__class_weights': [1, 8.87], 'catboost__depth': 10, 'catboost__l2_leaf_reg': 1, 'catboost__learning_rate': 0.03, 'tfidf__max_df': 0.25}
F1 лучшей модели -0.7312842301424006


## Модели проверим на тестовой выборке

In [25]:
%%time
count_tf_idf = TfidfVectorizer()

# преобразование TF-IDF для тестовой выборки
X_train_tfidf = count_tf_idf.fit_transform(X_train)

# преобразование TF-IDF для тестовой выборки
X_test_tfidf = count_tf_idf.transform(X_test)

# вывод размеров матриц TF-IDF
print("Размер X_train_tfidf:", X_train_tfidf.shape)
print("Размер X_test_tfidf:", X_test_tfidf.shape)



Размер X_train_tfidf: (127433, 129078)
Размер X_test_tfidf: (31859, 129078)
CPU times: user 5.44 s, sys: 2.3 ms, total: 5.44 s
Wall time: 4.59 s


In [26]:
#%%time

model = CatBoostClassifier(iterations=500, random_seed=42, depth = 10, l2_leaf_reg = 1, learning_rate = 0.03,
                           eval_metric='F1', task_type='GPU', class_weights = [1, 8.87])

model.fit(X_train_tfidf, y_train, eval_set=(X_test_tfidf, y_test), verbose=100, early_stopping_rounds=100)


y_pred = model.predict(X_test_tfidf)


f1 = f1_score(y_test, y_pred)
print("F1 Score: {:.2f}".format(f1))

0:	learn: 0.6058188	test: 0.5868049	best: 0.5868049 (0)	total: 341ms	remaining: 2m 50s
100:	learn: 0.8198479	test: 0.8014617	best: 0.8021232 (99)	total: 29.4s	remaining: 1m 55s
200:	learn: 0.8492386	test: 0.8300214	best: 0.8300214 (200)	total: 56.4s	remaining: 1m 23s
300:	learn: 0.8668895	test: 0.8442862	best: 0.8442862 (300)	total: 1m 22s	remaining: 54.8s
400:	learn: 0.8780812	test: 0.8521140	best: 0.8521140 (399)	total: 1m 49s	remaining: 27s
499:	learn: 0.8867957	test: 0.8570287	best: 0.8570287 (499)	total: 2m 15s	remaining: 0us
bestTest = 0.857028673
bestIteration = 499
F1 Score: 0.73


In [27]:
#%%time

model = LogisticRegression(C=15, random_state=42)

model.fit(X_train_tfidf, y_train)


y_pred = model.predict(X_test_tfidf)


f1 = f1_score(y_test, y_pred)
print("F1 Score: {:.2f}".format(f1))


F1 Score: 0.78


### Вывод

Лучше всего себя показала модель на тестовой выборке  LogisticRegression(C=15, random_state=42) с f1 мерой = 0.78

<div class="alert alert-info"><b>Комментарий ревьюера ✔️:</b> Попробовал с pipilin в Colab, правда google после трех прогонов падает, думаю только платная версия будет правильно работать. Правда надо ставить task_type='GPU', n_jobs = 1. Пока дошел до этого  -- сколько нервов. И проблема в том что на бесплатной версии все равно вылетает, просматривает три варианта и вылетает. И при установке в pipilin параметра     'tfidf__ngram_range': [(1, 1), (1, 2), (1, 3)], модель сразу вылетает при его перемене. </div>


<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Хорошо:)
    
    
P.S что такое 1+1?)
</div>

<div class="alert alert-info"><b> Осталось со старого проекта  </div>


## Bert

Стоит отметить, что модель обучена для работы с предложениями до 512 символов. 

In [28]:
%%time
1+1
# Использование токенизатора и модели `toxic-bert`
tokenizer = AutoTokenizer.from_pretrained('unitary/toxic-bert')
model = AutoModelForSequenceClassification.from_pretrained('unitary/toxic-bert')


CPU times: user 2.04 s, sys: 1.89 s, total: 3.93 s
Wall time: 11.6 s


<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Молодец, что взял токсичный берт
</div>

In [29]:
%%time
# код для подготовки данных
max_len = 128

tokenized = df['text'].apply(lambda x: tokenizer.encode(x, add_special_tokens=True, max_length=max_len, truncation=True))
padded = np.array([i + [0]*(max_len - len(i)) for i in tokenized.values])
attention_mask = np.where(padded != 0, 1, 0)

input_ids = torch.tensor(padded)  
attention_mask = torch.tensor(attention_mask)

dataset = TensorDataset(input_ids, attention_mask, torch.tensor(df['toxic'].values))

CPU times: user 34.2 s, sys: 299 ms, total: 34.5 s
Wall time: 34.2 s


In [30]:
%%time
# Подготовка DataLoader
batch_size = 32
data_loader = DataLoader(dataset, batch_size=batch_size)

CPU times: user 178 µs, sys: 30 µs, total: 208 µs
Wall time: 218 µs


In [31]:
%%time
# Выполнение прогнозирования
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model.to(device)
model.eval()
device

CPU times: user 1.33 s, sys: 666 ms, total: 1.99 s
Wall time: 5.16 s


device(type='cuda')

In [32]:
predictions = []

for batch in tqdm(data_loader):
    input_ids, attention_mask, _ = [b.to(device) for b in batch]

    with torch.no_grad(): 
        batch_predictions = model(input_ids, attention_mask=attention_mask)

    predictions.append(batch_predictions.logits[:, 0].cpu().numpy())

features = np.concatenate(predictions, axis=0)


  0%|          | 0/4978 [00:00<?, ?it/s]

In [33]:
threshold = 0.5
binary_predictions = (features > threshold).astype(int)


y_true = df['toxic'].values
f1 = f1_score(y_true, binary_predictions)
print("F1-мера:", f1)

F1-мера: 0.8312158492843016


<div class="alert alert-info"><b> Поменял на графическийб использовал, чуть подправил для сравнения с df['toxic']. Получилось намного быстрее.  </div>

> Indented block




<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Дааа, все верно:)

Давай воспользуемся `Google Colab` с `GPU`, так получится обработать все данные +- за 5-20 минут 
    
Сначала настроем среду:
    
+  `Среда выполнения` -> `сменить среду выполнения`    
+  Теперь выбираем `GPU` и `сохранить`    
    
Объявляем `BERT`:
    
```python
tokenizer = BertTokenizer.from_pretrained("unitary/toxic-bert")
model = BertModel.from_pretrained("unitary/toxic-bert")    
```

Далее создаем padded и attention_mask (максимальную длину возьмем 128 этого вполне должно хватить)
    
```python
from torch.nn.utils.rnn import pad_sequence

tokenized = df['text'].progress_apply(lambda x: tokenizer.encode(x, max_length=128, truncation=True, add_special_tokens=True)) #обрежет под нужное кол-во токенов

padded = pad_sequence([torch.as_tensor(seq) for seq in tokenized], batch_first=True) #добьет нулями  

attention_mask = padded > 0
attention_mask = attention_mask.type(torch.LongTensor)    
```
    
Далее нам нужно создать загрузчик и датасет
    
```python
from torch.utils.data import TensorDataset, DataLoader
dataset = TensorDataset(attention_mask, padded)
dataloader = DataLoader(dataset, batch_size=32, shuffle=False, num_workers=0) # создает загрузчик данных с батчом 32

device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu") #проверяет доступно ли нам GPU или нет
print(f'Device: {device}')
```
Про `DataLoader` и `TensorDataset`. Огромный плюс, в том, что тебе не придется самому делать батчи + удобно очень использовать и выгружать данные (на больших проектах очень выручают, поэтому будет круто если разберешься, там нет ничего сложного, если что могу помочь, если не получится) P.S если что, не обращай внимание на collate_fn
                
+  https://towardsdatascience.com/how-to-use-datasets-and-dataloader-in-pytorch-for-custom-text-data-270eed7f7c00    
    
 
Далее мы, чутка переписываем получения embeddings. Тут мы просто переводим батчи и модель на GPU в процессе выполнения, где и происходят все вычисления
    
```python
embeddings = []
model.to(device)
model.eval() 
for attention_mask, padded in notebook.tqdm(dataloader):
    attention_mask, padded = attention_mask.to(device), padded.to(device)
    
    with torch.no_grad(): 
        batch_embeddings = model(padded, attention_mask=attention_mask)
        
    embeddings.append(batch_embeddings[0][:,0,:].cpu().numpy()) 

features = np.concatenate(embeddings)       
```
    
P.S У меня пример для получения embedding, но у тебя будет аналогично:)
</div>



## Вывод

Мы построили инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 
Обучили модель классифицировать комментарии на позитивные и негативные.
Построил модель со значением метрики качества на тестовой выборке F1 = 0.78 для линейной регрессии и Взяли обученную модель Bert-toxic и построили на ней прогноз и измерили F1= 0.83
Использовали 3 способа построения определение токсичных коментарий:
1. нейроную сеть bert-toxic
2. модель машинного обучения catboost
3. модель машинного обучения линейной регрессии
У обоих моделей есть плюсы и минусы. 

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> 

+  Сделай предобработку для tfidf
+  Попробуй прогнать модельку на GPU (если нет времени, то главное проведи предобработку)
</div>

<div class="alert alert-info"><b> Поменял на графическийб использовал, чуть подправил для сравнения с df['toxic']. Получилось намного быстрее и более точный результат. Сделал предобработку.  </div>


<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> 

Для работы с текстами используют и другие подходы. Например, сейчас активно используются RNN (LSTM) и трансформеры (BERT и другие с улицы Сезам, например, ELMO). НО! Они не являются панацеей, не всегда они нужны, так как и TF-IDF или Word2Vec + модели из классического ML тоже могут справляться. \
BERT тяжелый, существует много его вариаций для разных задач, есть готовые модели, есть надстройки над библиотекой transformers. Если, обучать BERT на GPU (можно в Google Colab или Kaggle), то должно быть побыстрее.\
https://huggingface.co/transformers/model_doc/bert.html \
https://t.me/renat_alimbekov \
https://colah.github.io/posts/2015-08-Understanding-LSTMs/ - Про LSTM \
https://web.stanford.edu/~jurafsky/slp3/10.pdf - про энкодер-декодер модели, этеншены\
https://pytorch.org/tutorials/beginner/transformer_tutorial.html - официальный гайд
по трансформеру от создателей pytorch\
https://transformer.huggingface.co/ - поболтать с трансформером \
Библиотеки: allennlp, fairseq, transformers, tensorflow-text — множество реализованных
методов для трансформеров методов NLP \
Word2Vec https://radimrehurek.com/gensim/models/word2vec.html 
    
</div>

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Данные загружены и подготовлены
- [x]  Модели обучены
- [x]  Значение метрики *F1* не меньше 0.75
- [x]  Выводы написаны

<font color='blue'><b>Итоговый комментарий ревьюера</b></font>
<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>Алексей, получился хороший проект! 
    
Если есть  если есть какие либо вопросы я с удовольствием на них отвечу:) <br> Исправь, пожалуйста, замечания и жду проект на следующую проверку:) </div>
